pkg/lightning/tikv/local_sst_writer.go

```diff
@@ -0,0 +1,205 @@
+	writeStream, err := ingestClient.Write(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	u := uuid.New()
+	writeMeta := &import_sstpb.SSTMeta{
+		Uuid:        u[:],
+		RegionId:    r0.Meta.Id,
+		RegionEpoch: r0.Meta.RegionEpoch,
+		Range: &import_sstpb.Range{
+			Start: sortedKVs[0][0],
+			End:   sortedKVs[len(sortedKVs)-1][0],
+		},
+	}
+	rpcCtx := kvrpcpb.Context{
+		RegionId:    r0.Meta.Id,
+		RegionEpoch: r0.Meta.RegionEpoch,
+		Peer:        r0.Leader,
+	}
+	err = writeStream.Send(&import_sstpb.WriteRequest{
+		Chunk:   &import_sstpb.WriteRequest_Meta{Meta: writeMeta},
+		Context: &rpcCtx,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	batch := &import_sstpb.WriteBatch{
+		CommitTs: ts,
+		Pairs:    make([]*import_sstpb.Pair, 0, len(sortedKVs)),
+	}
+	for _, kv := range sortedKVs {
+		batch.Pairs = append(batch.Pairs, &import_sstpb.Pair{
+			Key:   kv[0],
+			Value: kv[1],
+		})
+	}
+	err = writeStream.Send(&import_sstpb.WriteRequest{
+		Chunk:   &import_sstpb.WriteRequest_Batch{Batch: batch},
+		Context: &rpcCtx,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	resp, err := writeStream.CloseAndRecv()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if resp.GetError() != nil {
+		return nil, errors.Errorf("write failed: %s", resp.GetError())
+	}
+	return resp.Metas, nil
+}
+
+// writeCFWriter generates SST files for TiKV's write column family.
+type writeCFWriter struct {
+	sstWriter *rockssst.Writer
+	ts        uint64
+}
+
+// newWriteCFWriter creates a new writeCFWriter. Currently `identity` is acquired
+// from a real SST file generated by TiKV.
+func newWriteCFWriter(
+	sstPath string,
+	ts uint64,
+	identity *rockssst.Identity,
+) (*writeCFWriter, error) {
+	f, err := vfs.Default.Create(sstPath)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	writable := objstorageprovider.NewFileWritable(f)
+	writer := rockssst.NewWriter(writable, rockssst.WriterOptions{
+		// TODO(lance6716): should read TiKV config to know compression algorithm.
+		Compression: rocks.ZstdCompression,
+		// TODO(lance6716): should check the behaviour is the exactly same.
+		FilterPolicy: rocksbloom.FilterPolicy(10),
+		MergerName:   "nullptr",
+		TablePropertyCollectors: []func() rockssst.TablePropertyCollector{
+			func() rockssst.TablePropertyCollector {
+				return newMVCCPropCollector(ts)
+			},
+			func() rockssst.TablePropertyCollector {
+				return newRangePropertiesCollector()
+			},
+			// titan is only triggered when SST compaction at TiKV side.
+			func() rockssst.TablePropertyCollector {
+				return mockCollector{name: "BlobFileSizeCollector"}
+			},
+		},
+	}, identity)
+	return &writeCFWriter{sstWriter: writer, ts: ts}, nil
+}
+
+// set mimic TiKV's `TxnSstWriter.put`. ref
+// https://github.com/tikv/tikv/blob/7793f1d5dc40206fe406ca001be1e0d7f1b83a8f/components/sst_importer/src/sst_writer.rs#L92.
+func (w *writeCFWriter) set(key, value []byte) error {
```

Comment:
can you write the structure of the KV inside comments , the refered tikv code has no comments too

Reply:
I added a more detailed TiKV link, please check if it's enough. The complete data layout is a bit complex and we only use part of flags in import code path.

Reply:
https://github.com/pingcap/tidb/pull/57061/commits/a5ac96f0504ba657322f2fe898184f425103ed07